primitives/state-machine/src/overlayed_changes/changeset.rs

```diff
@@ -0,0 +1,635 @@
+			}
+		} else {
+			Some(Some(init()))
+		};
+
+		if let Some(cloned) = clone_into_new_tx {
+			overlayed.set(cloned, first_write_in_tx, at_extrinsic);
+		}
+		overlayed.value_mut()
+	}
+
+	/// Set all values to deleted which are matched by the predicate.
+	///
+	/// Can be rolled back or committed when called inside a transaction.
+	pub fn clear_where(
+		&mut self,
+		predicate: impl Fn(&[u8], &OverlayedValue) -> bool,
+		at_extrinsic: Option<u32>
+	) {
+		for (key, val) in self.changes.iter_mut().filter(|(k, v)| predicate(k, v)) {
+			val.set(None, insert_dirty(&mut self.dirty_keys, key.to_owned()), at_extrinsic);
+		}
+	}
+
+	/// Get a list of all changes as seen by current transaction.
+	pub fn changes(&self) -> impl Iterator<Item=(&StorageKey, &OverlayedValue)> {
+		self.changes.iter()
+	}
+
+	/// Get the change that is next to the supplied key.
+	pub fn next_change(&self, key: &[u8]) -> Option<(&[u8], &OverlayedValue)> {
+		use std::ops::Bound;
+		let range = (Bound::Excluded(key), Bound::Unbounded);
+		self.changes.range::<[u8], _>(range).next().map(|(k, v)| (&k[..], v))
+	}
+
+	/// Consume this changeset and return all committed changes.
+	///
+	/// Panics:
+	/// Panics if there are open transactions: `transaction_depth() > 0`
+	pub fn drain_commited(self) -> impl Iterator<Item=(StorageKey, Option<StorageValue>)> {
+		assert!(self.transaction_depth() == 0, "Drain is not allowed with open transactions.");
+		self.changes.into_iter().map(|(k, mut v)| (k, v.pop_transaction().value))
+	}
+
+	/// Returns the current nesting depth of the transaction stack.
+	///
+	/// A value of zero means that no transaction is open and changes are committed on write.
+	pub fn transaction_depth(&self) -> usize {
+		self.dirty_keys.len()
+	}
+
+	/// Start a new nested transaction.
+	///
+	/// This allows to either commit or roll back all changes that where made while this
+	/// transaction was open. Any transaction must be closed by one of the aforementioned
+	/// functions before this overlay can be converted into storage changes.
+	///
+	/// Changes made without any open transaction are committed immediatly.
+	pub fn start_transaction(&mut self) {
+		self.dirty_keys.push(Default::default());
+	}
+
+	/// Rollback the last transaction started by `start_transaction`.
+	///
+	/// Any changes made during that transaction are discarded.
+	///
+	/// Panics:
+	/// Will panic if there is no open transaction.
+	pub fn rollback_transaction(&mut self) {
+		self.close_transaction(true);
+	}
+
+	/// Commit the last transaction started by `start_transaction`.
+	///
+	/// Any changes made during that transaction are committed.
+	///
+	/// Panics:
+	/// Will panic if there is no open transaction
+	pub fn commit_transaction(&mut self) {
+		self.close_transaction(false);
+	}
+
+	fn close_transaction(&mut self, rollback: bool) {
+		for key in self.dirty_keys.pop().expect(PROOF_DIRTY_KEYS) {
+			let value = self.changes.get_mut(&key).expect(PROOF_DIRTY_OVERLAY_VALUE);
```

Comment:
The overlayed value fetched from `self.changes` is called `overlayed` and here it is called `value`. I find it useful to use the same identifier for the same thing throughout the code to reduce mental overhead.